### ISSUE TITLE
Unexpected behaviour when handling default values with Environments

### ISSUE THREAD

**Issue by kgpayne** [meltano/meltano#6172]
### What is the current *bug* behavior?

Meltano does not take Environment inheritance into account when setting a value to its default in an Environment.

### What is the expected *correct* behavior?

Currently, when setting a setting value to it's default, Meltano will `unset` all values ("to fall back to default") rather than adding a new entry with the default value to `meltano.yml`. This behaviour was correct pre-environments, but does not take into account inheritance now that environments exist. The behaviour I would expect is that a new setting value would be added to the specific Environment with the default value, rather than the current behaviour that `unsets` values elsewhere in the inheritance chain.

### Steps to reproduce

```
# add example tap
meltano add extractor tap-gitlab

# configure example setting with non-default value in base
meltano --no-environment config tap-gitlab set fetch_merge_request_commits true

# configure same setting with default value in an Environment
meltano environment=dev config tap-gitlab set fetch_merge_request_commits false
```

The above results in the removal of the `True` setting in base, rather than the addition of a `False` setting value in the environment `dev`.

### Relevant logs and/or screenshots

<img width="1336" alt="Screenshot 2022-06-13 at 15 02 48" src="https://user-images.githubusercontent.com/5585874/173373187-e42fd497-2c58-4172-a0b7-7fa173c77fef.png">

Bug behaviour - removed `True` setting in plugin base configuration and no additional `False` setting in Environment `dev`.
Expected behaviour:

<img width="618" alt="Screenshot 2022-06-13 at 15 03 54" src="https://user-images.githubusercontent.com/5585874/173373791-2c20bdd0-6c21-4682-bf6d-2593e66f7daa.png">

`true` retained in base and `false` added to Environment `dev` (even though it happens to be the default value).

### Further regression test
_Ensure we automatically catch similar issues in the future_

- [ ] Write additional adequate test cases and submit test results
- [ ] Test results should be reviewed by a person from the team

/label ~bug
/label ~"To Do"

**Comment by DouweM** [comment]
@kgpayne Does this look related to https://github.com/meltano/meltano/issues/6116?

**Comment by DouweM** [comment]
@tayloramurphy I've put this one into the current milestone assuming you agree that fixing config/environments related bugs is a good use of our time right now!

**Comment by aaronsteers** [comment]
Unassigning, as this should be resolved by:

- #3419 
- #6116 
- #6247

**Comment by tayloramurphy** [comment]
@aaronsteers as the last issue in your comment is nearly merged, I'm assigning this to @kgpayne to validate the behavior is no longer present.

**Comment by kgpayne** [comment]
@tayloramurphy @aaronsteers I have confirmed the issue is still present in v2.3.0.

TLDR; we still need to fix this. Easiest way is to remove the trap door (and accept that users will in some cases store a copy of the default value in their `meltano.yml`). Harder way is make the trap door more intelligent, only calling `unset` if it is safe to do so (i.e. there are no other non-default values set elsewhere in other environments or the resolution hierarchy). Easy way is probably a 4, harder way is an 8 or 12 (I don't know that we have a precedent for checking values in all stores, including all Environments yet).

Root cause: This is still happening because we still have a [trap-door check](https://github.com/meltano/meltano/blob/1c4bec902f42f2c96957fdda1d2e1484cb5c065f/src/meltano/core/settings_store.py#L1287) in the `set` method on the `AutoStoreManager` which checks if the new value is equal to the default value, and if so calls `unset` to avoid storing a redundant copy of the default value in `meltano.yml`. This made sense pre-environments, but causes issues now in the case above where a non-default base configuration cannot be overridden by a default value in an Environment.

**Comment by tayloramurphy** [comment]
> Easiest way is to remove the trap door (and accept that users will in some cases store a copy of the default value in their meltano.yml).

I'd be in favor of the easier fix. This would solve the problem but leave the door open to possibly better solutions of making it more intelligent (maybe not needed) or bringing better UX and visibility to all the possible flavors of a setting. i.e. "Setting X is false in default, true in prod environment, false via env var" etc. 

@aaronsteers thoughts?

**Comment by aaronsteers** [comment]
Agreed I think the simply path is best, to simply remove the equivalence check and _always_ do a set when a set operation is requested.

I commented something similar in a related issue regarding the auto store detection. (Sorry, don't have that link handy.)
